Sources/IO/Core/HttpDataSetReader/index.js

```diff
@@ -145,26 +147,71 @@ function vtkHttpDataSetReader(publicAPI, model) {
   // Internal method to fetch Array
   function fetchArray(array, options = {}) {
     const arrayId = `${array.ref.id}|${array.vtkClass}`;
-    if (!cachedArrays[arrayId]) {
+
+    // Check if array is present in cache
+    if (model.maxCacheSize && cachedArrays[arrayId]) {
+      // Update last access for cache retention rules
+      cachedArrays[arrayId].lastAccess = new Date();
+      return Promise.resolve(cachedArrays[arrayId].array);
+    }
+
+    if (!cachedPromises[arrayId]) {
       // Cache the promise while fetching
-      cachedArrays[arrayId] = model.dataAccessHelper
+      cachedPromises[arrayId] = model.dataAccessHelper
         .fetchArray(publicAPI, model.baseURL, array, options)
         .then((newArray) => {
-          // Replace the promise with the array in cache once downloaded
-          cachedArrays[arrayId] = newArray;
+          // Remove cached promise
+          delete cachedPromises[arrayId];
+
+          // Return here if caching is disabled
+          if (!model.maxCacheSize) {
+            return newArray;
+          }
+
+          // Add array to cache
+          cachedArrays[arrayId] = { array: newArray, lastAccess: new Date() };
+
+          // If maxCacheSize is set to -1 the cache is unlimited
+          if (model.maxCacheSize === -1) {
+            return newArray;
+          }
+
+          // sort cache according to access times (potentially needed for creating space)
+          const sortedArrayCache = Object.entries(cachedArrays).sort((a, b) =>
+            Math.sign(b[1].lastAccess - a[1].lastAccess)
+          );
+
+          // Check cache size
+          const cacheSizeLimit = model.maxCacheSize * MiB;
+          let cacheSize = Object.values(cachedArrays).reduce(
+            (accSize, entry) => accSize + entry.array.values.byteLength,
+            0
+          );
```

Comment:
sort and reduce are not very fast, but I guess it is fine to leave it as is. I presume it is not on a critical path.

Reply:
I have tested it with 7000+ entries by letting the example I have attached to #2978 run for a while and there do not seem to be any issues. I leave it for now.